handlers/authentication.go

```diff
@@ -1,17 +1,214 @@
 package handlers
 
 import (
+	"fmt"
 	"net/http"
+	"strconv"
 	"strings"
+	"time"
 
+	"github.com/dgrijalva/jwt-go"
 	"github.com/kiali/kiali/business"
 	"github.com/kiali/kiali/config"
 	"github.com/kiali/kiali/log"
 )
 
 type AuthInfo struct {
-	Strategy              string `json:"strategy"`
-	AuthorizationEndpoint string `json:"authorizationEndpoint,omitempty"`
+	Strategy              string      `json:"strategy"`
+	AuthorizationEndpoint string      `json:"authorizationEndpoint,omitempty"`
+	SessionInfo           sessionInfo `json:"sessionInfo"`
+}
+
+type sessionInfo struct {
+	Username  string `json:"username,omitempty"`
+	ExpiresOn string `json:"expiresOn,omitempty"`
+}
+
+type TokenResponse struct {
+	Username  string `json:"username"`
+	Token     string `json:"token"`
+	ExpiresOn string `json:"expiresOn"`
+}
+
+func checkKialiCredentials(r *http.Request) string {
+	conf := config.Get()
+	u, p, ok := r.BasicAuth()
+
+	if ok && conf.Server.Credentials.Username == u && conf.Server.Credentials.Password == p {
+		return u
+	}
+
+	return ""
+}
+
+func getTokenStringFromRequest(r *http.Request) string {
+	tokenString := "" // Default to no token.
+
+	// Token can be provided in an authorization HTTP header
+	if headerValue := r.Header.Get("Authorization"); strings.Contains(headerValue, "Bearer") {
+		tokenString = strings.TrimPrefix(headerValue, "Bearer ")
+	}
+
+	// Token can also be provided by a browser in a Cookie.
+	// Checked at second. Thus, the token in the cookie has priority.
```

Comment:
I wonder if we could do this first and only check the header if we didn't have the cookie. Otherwise we are always checking twice.

Reply:
Done!